### PR TITLE
Return registration_endpoint in OidcProviderConfigurationEndpointFilter

### PR DESCRIPTION
**Pull request by sahariardev**
Before: client registration endpoint was not retuned in oidc
Provider Configuration response

After: Returns client registration endpoint in oidc provider configuration
response if client registration is enabled

Fixes gh-370

### CODE REVIEW

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java [hunk 1]
```diff
@@ -59,12 +59,18 @@ public final class OidcProviderConfigurationEndpointFilter extends OncePerReques
...
 	private final RequestMatcher requestMatcher;
+	private final boolean dynamicClientRegistrationEnabled;
 	private final OidcProviderConfigurationHttpMessageConverter providerConfigurationHttpMessageConverter =
 			new OidcProviderConfigurationHttpMessageConverter();
 
 	public OidcProviderConfigurationEndpointFilter(ProviderSettings providerSettings) {
+		this(providerSettings, false);
+	}
+
+	public OidcProviderConfigurationEndpointFilter(ProviderSettings providerSettings, boolean dynamicClientRegistrationEnabled) {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java:70]: Adding the flag `dynamicClientRegistrationEnabled` to the constructor will not provide us the flexibility we need. There will be other features that are enabled/disabled and adding more flags to constructor provides limited flexibility and API will constantly change.

Comment (sahariardev) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java:70]: Hi @jgrandja,
Thanks for your feedback.
One thing we can do is keep this flag inside provider settings as this is a provider related settings. What do you think? Or do you have any better idea for this keeping this flag?

Comment (sahariardev) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java:70]: Thanks for your feed, will try to find a way out

Comment (sahariardev) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java]: Hi @jgrandja,
I have updated the PR. Please have a look and provider you feedback.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java [hunk 2]
```diff
@@ -119,4 +127,17 @@ private static Consumer<List<String>> clientAuthenticationMethods() {
 	private static String asUrl(String issuer, String endpoint) {
 		return UriComponentsBuilder.fromUriString(issuer).path(endpoint).build().toUriString();
 	}
+
+	private boolean isDynamicClientRegistrationEnabled() {
+		DefaultSecurityFilterChain defaultSecurityFilterChain = (DefaultSecurityFilterChain) WebApplicationContextUtils
+				.getRequiredWebApplicationContext(getServletContext()).getBean("authorizationServerSecurityFilterChain");
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java:133]: Thanks for the updates @sahariardev.

We cannot rely on the bean being called `authorizationServerSecurityFilterChain`. The consuming application could name the bean to something different and then this logic would not work.

Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java]: @sahariardev Sorry for the delayed response. I haven't had a chance to revisit this yet. I have some re-factoring tasks that are higher priority at the moment. I will touch base as soon as I get a chance to revisit this.

Comment (sahariardev) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java]: Okay, no problem.

Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java]: Thanks for your patience @sahariardev. I just merged gh-616 via 0994a1e1e14263d106373e3397f8c5c44f9a7708.

Based on the updates in that commit, the following code (add to `OidcConfigurer.configure()`) will allow you to determine when `oidc.clientRegistrationEndpoint()` is enabled so you can apply the `registration_endpoint` configuration:

```java
OidcClientRegistrationEndpointConfigurer clientRegistrationEndpointConfigurer =
		getConfigurer(OidcClientRegistrationEndpointConfigurer.class);
if (clientRegistrationEndpointConfigurer != null) {
	OidcProviderConfigurationEndpointConfigurer providerConfigurationEndpointConfigurer =
			getConfigurer(OidcProviderConfigurationEndpointConfigurer.class);
	providerConfigurationEndpointConfigurer.addDefaultProviderConfigurationCustomizer(
			builder -> builder.clientRegistrationEndpoint("https://localhost/connect/register"));
}
```

Let me know if you have any questions. Thanks.

Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java]: Also, please rebase the PR off of `0.4.x` branch.

Comment (sahariardev) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java]: Thanks @jgrandja. I will update those ASAP.

Comment (sahariardev) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java]: Hi  @jgrandja,
I have started working on this. As I was facing some trouble to rebase with 0.4.x, I had do close this one and opening a new PR. Will let you know the updates ASAP.

Comment (sahariardev) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcProviderConfigurationEndpointFilter.java]: Here is the new PR link. let's continue our discussion there.

**Pull request closed without merging**